### PR TITLE
[RHELC-778] Add progress indicators for YUM and DNF modules

### PR DESCRIPTION
**Pull request by r0x0d**
Add progress indicators for DNF/YUM

This commit introduces the callbacks classes for the progress indicators
that we will be showing for the user during a few different phases in
our code.

The callbacks will called during these phases:

* For DNF: Dependency Solver, Package Download and Transaction Progress
* For YUM: Package Download and Transaction Progress

We don't have a dependency solver callback for yum, because the yum code
base does not provide any base class for us to override and iterate
over, thus, we would require to copy and paste their own dependency
resolver callback and deal adjust the code with the things we want and
the ones we don't care. Since those classes are highly tied to the
`base` usage, it is possible that we could mess with something and lead
to obscure errors, thus, we are not using any dependency solver callback
for YUM.

Signed-off-by: Rodolfo Olivieri <rolivier@redhat.com>

Jira Issue: [RHELC-778](https://issues.redhat.com/browse/RHELC-778)

Checklist
- [ ] PR meets acceptance criteria specified in the Jira issue
- [ ] PR has been tested manually in a VM (either author or reviewer)
- [ ] Jira issue has been made public if possible
- [x] `[RHELC-]` is part of the PR title 
- [ ] Code and tests are documented properly
- [ ] The commits are squashed to as few commits as possible (without losing data) 
- [ ] When merged: Jira issue has been updated to `Release Pending`

### CODE REVIEW

File: convert2rhel/pkgmanager/handlers/dnf/callback.py
Comment (r0x0d) [reply convert2rhel/pkgmanager/handlers/dnf/callback.py]: fixed in https://github.com/oamg/convert2rhel/pull/709/commits/cb637aee045677ccd2aebfeb877b5f77e5f23190

**Comment by r0x0d** [comment]
@abadger now pre-commit.ci is passing again!

**Comment by danmyway** [comment]
On a call with @r0x0d and @kokesak, we have agreed that verification for this feature would be more or less sanity check. The nature of the implemented changes is cosmetic, therefore should not break anything inside the conversion. The current test suite is sufficient to catch any failures.
We have also agreed on, that we will implement a sanity check in the future, which will match a list of strings in the log file after conversion. This feature output should be also implemented in said sanity check.

File: convert2rhel/pkgmanager/handlers/yum/callback.py
Comment (SpyTec) [inline convert2rhel/pkgmanager/handlers/yum/callback.py:140]:
```suggestion
        :param ts_current: Number, in order, of the current transaction in the transaction set.
        :type ts_current: int
```

File: convert2rhel/pkgmanager/handlers/dnf/__init__.py
Comment (SpyTec) [inline convert2rhel/pkgmanager/handlers/dnf/__init__.py:131]: A lot of the log changes prepend a newline, I've not run it myself to check but is it necessarily better?

**Review by SpyTec — COMMENTED** [review]
Not tested it in VM

File: convert2rhel/pkgmanager/handlers/dnf/__init__.py
Comment (r0x0d) [reply convert2rhel/pkgmanager/handlers/dnf/__init__.py:131]: It is easier to follow the output on the terminal. Before, every line was below each other, which, in my opinion, make this things a little harder to see. 

The idea here is to separate by "concerns" the logs, so the depsolver will be together, and the download will be separate from the depsolver, but the logs that follow the packages download will be together too and so on...

File: convert2rhel/pkgmanager/handlers/dnf/callback.py
Comment (r0x0d) [reply convert2rhel/pkgmanager/handlers/dnf/callback.py:256]: The base API will call this class on every package update, no matter if it is the same or not. See below an example: 

```bash
Reinstalling: pcre-8.42-6.el8.x86_64 [124/1208]
Reinstalling: pcre-8.42-6.el8.x86_64 [124/1208]
Reinstalling: pcre-8.42-6.el8.x86_64 [124/1208]
Reinstalling: pcre-8.42-6.el8.x86_64 [124/1208]
Reinstalling: pcre-8.42-6.el8.x86_64 [124/1208]
Reinstalling: pcre-8.42-6.el8.x86_64 [124/1208]
Reinstalling: pcre-8.42-6.el8.x86_64 [124/1208]
Reinstalling: pcre-8.42-6.el8.x86_64 [124/1208]
Reinstalling: pcre-8.42-6.el8.x86_64 [124/1208]
Reinstalling: pcre-8.42-6.el8.x86_64 [124/1208]
Reinstalling: pcre-8.42-6.el8.x86_64 [124/1208]
Reinstalling: pcre-8.42-6.el8.x86_64 [124/1208]
Reinstalling: pcre-8.42-6.el8.x86_64 [124/1208]
Reinstalling: pcre-8.42-6.el8.x86_64 [124/1208]
Reinstalling: pcre-8.42-6.el8.x86_64 [124/1208]
Reinstalling: pcre-8.42-6.el8.x86_64 [124/1208]
Reinstalling: pcre-8.42-6.el8.x86_64 [124/1208]
Reinstalling: pcre-8.42-6.el8.x86_64 [124/1208]
Reinstalling: pcre-8.42-6.el8.x86_64 [124/1208]
Reinstalling: pcre-8.42-6.el8.x86_64 [124/1208]
Reinstalling: pcre-8.42-6.el8.x86_64 [124/1208]
Reinstalling: pcre-8.42-6.el8.x86_64 [124/1208]
Reinstalling: pcre-8.42-6.el8.x86_64 [124/1208]
Reinstalling: pcre-8.42-6.el8.x86_64 [124/1208]
Reinstalling: pcre-8.42-6.el8.x86_64 [124/1208]
Reinstalling: pcre-8.42-6.el8.x86_64 [124/1208]
Reinstalling: pcre-8.42-6.el8.x86_64 [124/1208]
Reinstalling: pcre-8.42-6.el8.x86_64 [124/1208]
Reinstalling: pcre-8.42-6.el8.x86_64 [124/1208]
Reinstalling: pcre-8.42-6.el8.x86_64 [124/1208]
Reinstalling: pcre-8.42-6.el8.x86_64 [124/1208]
```

Comment (r0x0d) [reply convert2rhel/pkgmanager/handlers/dnf/callback.py:256]: I will update the comment with a more clear message.

File: convert2rhel/pkgmanager/handlers/dnf/__init__.py
Comment (abadger) [reply convert2rhel/pkgmanager/handlers/dnf/__init__.py:134]: I think stylistically, logs should generally be line oriented, one line per entry but what you say makes perfect sense for terminal output.  I'm not sure if we can get both behaviours with the way we use logger for displaying the things we want the user to read....  

Maybe we could write some code in our custom logger to strip leading and trailing newlines from the message just when outputting to a log file, but not to the terminal?  Or we could create a set of functions that will output strings to the terminal only (It is a little more than simply calling print() because we would only want it to run if the logging level is appropriate).

File: MANIFEST.in
Comment (r0x0d) [reply MANIFEST.in:6]: Well, this file mainly contains overrides for the tools that we use during development. I thought that this was not necessary in the tarball, as it is not really being used by anything, specially because we haven't migrated to use only pyproject.toml with setuptools and drop the `setup.{cfg,py}` ones.

File: convert2rhel/pkgmanager/handlers/dnf/__init__.py
Comment (SpyTec) [reply convert2rhel/pkgmanager/handlers/dnf/__init__.py:134]: I think that might be a lot of extra work. So I vote for us reverting this and going with per-line logs rather than empty lines

Comment (r0x0d) [reply convert2rhel/pkgmanager/handlers/dnf/__init__.py:134]: No worries. Will revert the change in the next push.

File: MANIFEST.in
Comment (r0x0d) [reply MANIFEST.in:6]: does that make sense to you, @abadger?

Comment (abadger) [reply MANIFEST.in:6]: I think I would include pyproject.toml if the only issue is whether it is necessary.  On a practical side, I know that packagers like to run the tests that upstreams have produced so leaving out some of the configuration for tools that we use in our pre-commit configuration doesn't seem right (Although packagers *probably* will only run pytest and not something like `pre-commit run --all-files`).

I also know that the GPL requires corresponding source which includes all the "build files" required to transform the source.  With configuration of developer tools for an interpreted language, the line is not as clear as with compiled code but I like to err on the side of caution there.

File: convert2rhel/pkgmanager/__init__.py
Comment (r0x0d) [inline convert2rhel/pkgmanager/__init__.py:18]: Another future note for me: This might get merge conflicts if https://github.com/oamg/convert2rhel/pull/713 is merged first.

File: convert2rhel/pkgmanager/handlers/dnf/callback.py
Comment (abadger) [inline convert2rhel/pkgmanager/handlers/dnf/callback.py:196]:
```suggestion
            if status == pkgmanager.callback.STATUS_DRPM and self.total_drpm > 1:
```

File: convert2rhel/pkgmanager/handlers/dnf/callback.py
Comment (abadger) [inline convert2rhel/pkgmanager/handlers/dnf/callback.py:90]:
```suggestion
        This method sends a message that indicates what the package added
        to the dependency solver transaction will be doing.
```

File: convert2rhel/pkgmanager/handlers/dnf/callback.py
Comment (abadger) [inline convert2rhel/pkgmanager/handlers/dnf/callback.py:92]:
```suggestion
        :param pkg: The package itself added to the dependency solver. It's
```

File: convert2rhel/pkgmanager/handlers/dnf/callback.py
Comment (abadger) [inline convert2rhel/pkgmanager/handlers/dnf/callback.py:263]:
```suggestion
        # no matter if it is the same update or not, so, the below statement
```

File: convert2rhel/pkgmanager/handlers/dnf/callback.py
Comment (abadger) [inline convert2rhel/pkgmanager/handlers/dnf/callback.py:274]:
```suggestion
            If there is no problem reported during the scriptlet execution,
```

File: convert2rhel/pkgmanager/handlers/dnf/callback.py
Comment (abadger) [inline convert2rhel/pkgmanager/handlers/dnf/callback.py:281]:
```suggestion
        # no matter if the messages are empty or not, so, the below statement
```

File: convert2rhel/pkgmanager/handlers/yum/callback.py
Comment (abadger) [inline convert2rhel/pkgmanager/handlers/yum/callback.py:90]:
```suggestion
        which were downloaded, no need to do any fancy progress-bar.
```

File: convert2rhel/pkgmanager/handlers/yum/callback.py
Comment (abadger) [inline convert2rhel/pkgmanager/handlers/yum/callback.py:95]:
```suggestion
        :param frac: A number between `0` and `1` that represents the percentage of the file that has been downloaded.
        :type frac: float
```
From the first part of the parameter description, I think this is what should be here but feel free to correct me if I'm wrong.

File: convert2rhel/pkgmanager/handlers/yum/callback.py
Comment (abadger) [inline convert2rhel/pkgmanager/handlers/yum/callback.py:96]:
```suggestion
        :param fread: String containing how many bytes were read in the download.
```

File: convert2rhel/pkgmanager/handlers/yum/callback.py
Comment (abadger) [inline convert2rhel/pkgmanager/handlers/yum/callback.py:98]:
```suggestion
        :param ftime: String that contains the remaining time for the package to be finished downloading.
```

File: convert2rhel/pkgmanager/handlers/yum/callback.py
Comment (abadger) [inline convert2rhel/pkgmanager/handlers/yum/callback.py:128]:
```suggestion
            Yum API sends the `package` paramter as two different types.
```

File: convert2rhel/pkgmanager/handlers/yum/callback.py
Comment (abadger) [inline convert2rhel/pkgmanager/handlers/yum/callback.py:138]:
```suggestion
        :param te_current: How much work has already been done in the transaction.
```

File: convert2rhel/pkgmanager/handlers/yum/callback.py
Comment (abadger) [inline convert2rhel/pkgmanager/handlers/yum/callback.py:144]:
```suggestion
        :param ts_total: How many transactions are present in the transaction set.
```

File: convert2rhel/pkgmanager/handlers/yum/callback.py
Comment (abadger) [inline convert2rhel/pkgmanager/handlers/yum/callback.py:152]:
```suggestion
        # Here we convert the package to a str because we just need to do a
        # standard str comparision rather than what is implemented in
        # `yum.sqlitesack.YumAvailablePackageSqlite` (that class does
        # version comparison and a bunch of other stuff). We don't care about
```

File: convert2rhel/pkgmanager/handlers/yum/callback.py
Comment (abadger) [inline convert2rhel/pkgmanager/handlers/yum/callback.py:168]:
```suggestion
        """Hook for reporting output from an rpm scriptlet.
```

File: convert2rhel/pkgmanager/handlers/yum/callback.py
Comment (abadger) [inline convert2rhel/pkgmanager/handlers/yum/callback.py:171]:
```suggestion
            If there is no problem reported during the scriptlet execution,
```

File: convert2rhel/pkgmanager/handlers/yum/callback.py
Comment (abadger) [inline convert2rhel/pkgmanager/handlers/yum/callback.py:180]:
```suggestion
        # no matter if the messages are empty or not, so, the below statement
```

File: convert2rhel/pkgmanager/handlers/dnf/callback.py
Comment (abadger) [inline convert2rhel/pkgmanager/handlers/dnf/callback.py:147]:
```suggestion
        progress of the downloaded packages. Without them, we lose the
        information about how many packages need to be downloaded, and where
```

File: convert2rhel/pkgmanager/handlers/dnf/callback.py
Comment (abadger) [inline convert2rhel/pkgmanager/handlers/dnf/callback.py:171]:
```suggestion
            DNF will give us status about the file that is being downloaded, so
            we can print this status to the user, for them to know that
```

File: convert2rhel/pkgmanager/handlers/dnf/callback.py
Comment (abadger) [inline convert2rhel/pkgmanager/handlers/dnf/callback.py:255]:
```suggestion
        # Here we convert the package to a str because we just need to do a
        # standard str comparision rather than what is implemented in
        # `yum.sqlitesack.YumAvailablePackageSqlite` (that class does
        # version comparison and a bunch of other stuff). We don't care about
```

**Review by abadger — APPROVED** [review]
One code change (that is very minor) and some grammar and spelling changes which aren't needed.  This looks good to me.  Note that I have not tested the change in a VM.  Approving if you have tested.

File: convert2rhel/pkgmanager/handlers/yum/callback.py
Comment (pr-watson) [inline convert2rhel/pkgmanager/handlers/yum/callback.py:105]: change "abut" to "about"

Comment (r0x0d) [inline convert2rhel/pkgmanager/handlers/yum/callback.py:105]:
```suggestion
            # Metadata download about repositories will be sent to this class too.
```

Comment (r0x0d) [reply convert2rhel/pkgmanager/handlers/yum/callback.py]: done!

File: convert2rhel/pkgmanager/handlers/yum/callback.py
Comment (pr-watson) [inline convert2rhel/pkgmanager/handlers/yum/callback.py:128]:
```suggestion
            Yum API sends the `package` parameter as two different types.
```

**Review by SpyTec — APPROVED** [review]
Overall seems good to me not tested it in VM. If you can demo this next week on the demo session on Tuesday that would be great @r0x0d

**Review by pr-watson — APPROVED** [review]
These changes look good to me, approved

**Review by abadger — DISMISSED** [review]

**Comment by r0x0d** [comment]
> Overall seems good to me not tested it in VM. If you can demo this next week on the demo session on Tuesday that would be great @r0x0d

Nice idea! I will add that to the document

**Comment by r0x0d** [comment]
Woops, file conflicts... I will do ahead and fix them.

**Comment by r0x0d** [comment]
Sorry to who approved the PR.. It will need another approval :D

**Review by pr-watson — APPROVED** [review]

**Comment by r0x0d** [comment]
Pipeline is all green! Merging this as I already had approval before of the rebase.

**Pull request merged**
